spring-kafka/src/main/java/org/springframework/kafka/listener/TimestampedException.java

```diff
@@ -34,14 +34,36 @@
 
 	private final long timestamp;
 
+	@Deprecated
 	public TimestampedException(Exception ex, Clock clock) {
```

Comment:
I won't mind to have this deprecated, but I wonder what drove you to use `Clock` as a required from end-user API in the beginning?

I never saw this class usage in the wild, but I see it more and more in your code 😄 

So, what is really a benefit to of this class and how does it help being a required dependency in the API for end-users?

Thanks

Reply:
Well, having `Clock` passed as a constructor parameter enables us to unit test time-sensitive logic, by injecting mock clocks, so that's the main driver. AFAIK, it's considered a good practice over methods such as System.currentTimeMillis(), which you can't really precisely unit test. 

The other reason would be enabling users to setup any clock they want, maybe there's more than one for different timezones, and whatnot. I've worked with applications that defined their own clocks - never more than one though.

So I've seen and used `Clock` for this in application code quite a few times for this reason. You can see [this SO question](https://stackoverflow.com/questions/27067049/unit-testing-a-class-with-a-java-8-clock) to confirm I'm not alone 😄

Also, for example, in Apache Kafka you have the [Time](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/Time.java) interface, which is used all around for the same reason - just to give a couple examples, there's one in [KafkaConsumer.java](https://github.com/apache/kafka/blob/206ad4d2b530b1c0c776629256352cd044e68197/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L829-L837), or the [TimestampedWindowStoreBuilder.java](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/TimestampedWindowStoreBuilder.java) class.

So that's to say I didn't really invent this 😄 But I also think I may have gone too far in adding this to an `Exception`, so that's partially why I backed out on this.

Reply:
Thanks, @tomazfernandes .

I will need to learn more about this `Clock` abstraction.

My old school dictates me that no any testing hooks should be leaking into a production code, so we ended up with many reflection in our tests to modify a timestamp for assertions and so on 
Perhaps it's time to revise my vision for testing approaches 😄